packages/contracts-bedrock/src/L2/L1Block.sol

```diff
@@ -116,4 +126,86 @@ contract L1Block is ISemver {
+    /// Params are expected to be in the following order:
+    ///   1. _baseFeeScalar      L1 base fee scalar
+    ///   2. _blobBaseFeeScalar  L1 blob base fee scalar
+    ///   3. _sequenceNumber     Number of L2 blocks since epoch start.
+    ///   4. _timestamp          L1 timestamp.
+    ///   5. _number             L1 blocknumber.
+    ///   6. _basefee            L1 base fee.
+    ///   7. _blobBaseFee        L1 blob base fee.
+    ///   8. _hash               L1 blockhash.
+    ///   9. _batcherHash        Versioned hash to authenticate batcher by.
+    ///  10. _dependencySetSize  Size of the interop dependency set.
+    ///  11. _dependencySet      Array of chain IDs for the interop dependency set.
+    function setL1BlockValuesInterop() external {
+        bytes4 errorNotDepositorSelector = NotDepositor.selector;
+        bytes4 errorDependencySetSizeMismatch = DependencySetSizeMismatch.selector;
+        assembly {
+            // Revert if the caller is not the depositor account.
+            if xor(caller(), DEPOSITOR_ACCOUNT) {
+                mstore(0x00, errorNotDepositorSelector)
+                revert(0x00, 4) // returns the stored 4-byte selector from above
+            }
+            // sequencenum (uint64), blobBaseFeeScalar (uint32), baseFeeScalar (uint32)
+            sstore(sequenceNumber.slot, shr(128, calldataload(4)))
+            // number (uint64) and timestamp (uint64)
+            sstore(number.slot, shr(128, calldataload(20)))
+            sstore(basefee.slot, calldataload(36)) // uint256
+            sstore(blobBaseFee.slot, calldataload(68)) // uint256
+            sstore(hash.slot, calldataload(100)) // bytes32
+            sstore(batcherHash.slot, calldataload(132)) // bytes32
+
+            // Load dependencySetSize from calldata (at offset 164 after calldata for setL1BlockValuesEcotone ends)
+            let dependencySetSize_ := shr(248, calldataload(164))
+
+            // Revert if dependencySetSize_ doesn't match the length of dependencySet in calldata
+            if xor(add(165, mul(dependencySetSize_, 0x20)), calldatasize()) {
+                mstore(0x00, errorDependencySetSizeMismatch)
+                revert(0x00, 4) // returns the stored 4-byte selector from above
+            }
+
+            // Use memory to hash and get the start index of dependencySet
+            mstore(0x00, dependencySet.slot)
+            let dependencySetStartIndex := keccak256(0x00, 0x20)
+
+            // Iterate over calldata dependencySet and write to store dependencySet
+            for { let i := 0 } lt(i, dependencySetSize_) { i := add(i, 1) } {
+                // Increase length of dependencySet array
+                sstore(dependencySet.slot, add(sload(dependencySet.slot), 1))
+
+                // Load value from calldata and write to storage (dependencySet) at index
+                let val := calldataload(add(165, mul(i, 0x20)))
+                sstore(add(dependencySetStartIndex, i), val)
```

Comment:
This isn't right - this is how storage is done for fixed size arrays but not for dynamic sized arrays. This makes adding any new storage slot unsafe. For dynamic sized arrays, see the docs [here](https://docs.soliditylang.org/en/v0.8.25/internals/layout_in_storage.html#mappings-and-dynamic-arrays). We will need to have an additional hash before the `sstore` - otherwise we cannot add storage variables after the `dependencySetStartIndex` unless its padded by `type(uint8).max`

Reply:
can you elaborate more? `dependencySetStartIndex` is the hash of the storage slot for dependencySet

https://github.com/ethereum-optimism/optimism/blob/c7722f50622766a9331ef99cd1ad50bae31c48ba/packages/contracts-bedrock/src/L2/L1Block.sol#L171-L173


which seems to make sense, given the [docs](https://docs.soliditylang.org/en/v0.8.25/internals/layout_in_storage.html#mappings-and-dynamic-arrays): "Assume the storage location of the mapping or array ends up being a slot p after applying the storage layout rules. For dynamic arrays, this slot stores the number of elements in the array. [...] Array data is located starting at keccak256(p) and it is laid out in the same way as statically-sized array data would: One element after the other [...]"

Reply:
Ok I was thinking of how strings are stored in storage, they are not in storage slots right next to each other but it does appear that you are correct with arrays. The getter test would fail if this was not correct